<!DOCTYPE html>
<html  lang="zh">
<head>
    <meta charset="utf-8" />

<meta name="generator" content="Hexo 4.2.0" />

<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />

<title>【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记 - Robin唔系肉饼</title>


    <meta name="description" content="*SingleR*是用于单细胞RNA测序（scRNAseq）数据的自动注释方法  一、内置数据集下载与加载 用函数下载参考数据集，并返回SummarizedExperiment对象，具有“ logcounts” assay，其中包含对数标准化的表达值以及colData中的细胞类型标签。  人类参考集 从Blueprint和ENCODE获取人类bulk RNA-seq数据。 1BlueprintE">
<meta property="og:type" content="article">
<meta property="og:title" content="【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记">
<meta property="og:url" content="http://yoursite.com/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%848%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E5%8C%85%E7%9A%84%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/index.html">
<meta property="og:site_name" content="Robin唔系肉饼">
<meta property="og:description" content="*SingleR*是用于单细胞RNA测序（scRNAseq）数据的自动注释方法  一、内置数据集下载与加载 用函数下载参考数据集，并返回SummarizedExperiment对象，具有“ logcounts” assay，其中包含对数标准化的表达值以及colData中的细胞类型标签。  人类参考集 从Blueprint和ENCODE获取人类bulk RNA-seq数据。 1BlueprintE">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://yoursite.com/images/og_image.png">
<meta property="article:published_time" content="2020-02-06T08:17:54.000Z">
<meta property="article:modified_time" content="2020-02-06T08:18:56.247Z">
<meta property="article:author" content="Robin Hood">
<meta property="article:tag" content="学习与分享">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yoursite.com/images/og_image.png">







<link rel="icon" href="/images/favicon.svg">


<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma@0.7.2/css/bulma.css">
<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css">
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:400,600|Source+Code+Pro">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@9.12.0/styles/atom-one-light.css">


    
    
<style>body>.footer,body>.navbar,body>.section{opacity:0}</style>

    
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lightgallery@1.6.8/dist/css/lightgallery.min.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/justifiedGallery@3.7.0/dist/css/justifiedGallery.min.css">

    
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/outdatedbrowser@1.1.5/outdatedbrowser/outdatedbrowser.min.css">

    
    
    
    
<link rel="stylesheet" href="/css/back-to-top.css">

    
    
    
    
    
    
    
    <link rel="stylesheet" href="/css/progressbar.css">
<script src="https://cdn.jsdelivr.net/npm/pace-js@1.0.2/pace.min.js"></script>
    
    
    


<link rel="stylesheet" href="/css/style.css">
</head>
<body class="is-3-column">
    <nav class="navbar navbar-main">
    <div class="container">
        <div class="navbar-brand is-flex-center">
            <a class="navbar-item navbar-logo" href="/">
            
                <img src="/images/logo.svg" alt="【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记" height="28">
            
            </a>
        </div>
        <div class="navbar-menu">
            
            <div class="navbar-start">
                
                <a class="navbar-item"
                href="/">Home</a>
                
                <a class="navbar-item"
                href="/archives">Archives</a>
                
                <a class="navbar-item"
                href="/categories">Categories</a>
                
                <a class="navbar-item"
                href="/tags">Tags</a>
                
                <a class="navbar-item"
                href="/about">About</a>
                
            </div>
            
            <div class="navbar-end">
                
                    
                    <a class="navbar-item" target="_blank" rel="noopener" title="Download on GitHub" href="https://github.com/ppoffice/hexo-theme-icarus">
                        
                        <i class="fab fa-github"></i>
                        
                    </a>
                    
                
                
                
                <a class="navbar-item search" title="搜索" href="javascript:;">
                    <i class="fas fa-search"></i>
                </a>
                
            </div>
        </div>
    </div>
</nav>
    
    <section class="section">
        <div class="container">
            <div class="columns">
                <div class="column is-8-tablet is-8-desktop is-6-widescreen has-order-2 column-main">
<div class="card">
    
    <div class="card-content article ">
        
        <div class="level article-meta is-size-7 is-uppercase is-mobile is-overflow-x-auto">
            <div class="level-left">
                <time class="level-item has-text-grey" datetime="2020-02-06T08:17:54.000Z">2020-02-06</time>
                
                
                <span class="level-item has-text-grey">
                    
                    
                    39 分钟 读完 (大约 5777 个字)
                </span>
                
                
            </div>
        </div>
        
        <h1 class="title is-size-3 is-size-4-mobile has-text-weight-normal">
            
                【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记
            
        </h1>
        <div class="content">
            <blockquote>
<p>*<a href="https://bioconductor.org/packages/3.11/SingleR">SingleR</a>*是用于单细胞RNA测序（scRNAseq）数据的自动注释方法</p>
</blockquote>
<h1 id="一、内置数据集下载与加载"><a href="#一、内置数据集下载与加载" class="headerlink" title="一、内置数据集下载与加载"></a>一、内置数据集下载与加载</h1><blockquote>
<p>用函数下载参考数据集，并返回SummarizedExperiment对象，具有“ logcounts” assay，其中包含对数标准化的表达值以及colData中的细胞类型标签。</p>
</blockquote>
<h2 id="人类参考集"><a href="#人类参考集" class="headerlink" title="人类参考集"></a>人类参考集</h2><ol>
<li><p>从Blueprint和ENCODE获取人类bulk RNA-seq数据。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">BlueprintEncodeData(rm.NA &#x3D; c(&quot;rows&quot;, &quot;cols&quot;, &quot;both&quot;, &quot;none&quot;))</span><br></pre></td></tr></table></figure>
<blockquote>
<ul>
<li><strong>rm.NA</strong>：字符串型，指定应如何处理缺失值。 “rows”将删除具有至少一个缺失值的基因，“ cols”将删除具有至少一个缺失值的样本，“both”将删除任何具有至少一个缺失值的基因或样本，而“none”将不执行任何 去除。</li>
</ul>
</blockquote>
</li>
<li><p>从免疫细胞表达数据库（DICE）下载并缓存已分类细胞群的1561个bulk RNA-seq</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- DatabaseImmuneCellExpressionData()</span><br></pre></td></tr></table></figure></li>
<li><p>获得HPCA数据。下载并缓存那些存储在人类原代细胞图集Human Primary<br>Cell Atlas中的数据的标准化表达值。 </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- HumanPrimaryCellAtlasData()</span><br></pre></td></tr></table></figure>
<ol start="4">
<li>获得用于分类的造血细胞的bulk微阵列表达。下载并缓存可在GSE24759中找到的211个分类的造血细胞群体的bulk人类微阵列样品的标准化表达值<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- NovershternHematopoieticData()</span><br></pre></td></tr></table></figure></li>
<li>获取分类的人类免疫细胞的大量RNA-seq数据。下载并缓存可在GSE107011中找到的分类免疫细胞群的114个bulk RNA-seq样本的标准化表达值。<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- MonacoImmuneData()</span><br></pre></td></tr></table></figure>
<h2 id="小鼠参考集"><a href="#小鼠参考集" class="headerlink" title="小鼠参考集"></a>小鼠参考集</h2></li>
</ol>
</li>
<li><p>从免疫基因组计划中获取小鼠bulk表达数据。下载并缓存由免疫基因组计划（ImmGen）生成的830个纯小鼠免疫细胞微阵列样品的标准化表达值。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- ImmGenData()</span><br></pre></td></tr></table></figure>
<ol start="2">
<li>获取分类细胞群体的小鼠bulk表达数据（RNA-seq）。下载并缓存可在GEO处找到的358个分类细胞群体的bulk RNA-seq样本的标准化表达值。<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- MouseRNAseqData()</span><br></pre></td></tr></table></figure>

</li>
</ol>
</li>
</ol>
<p>================================================================</p>
<h1 id="二、-aggregateReference"><a href="#二、-aggregateReference" class="headerlink" title="二、 aggregateReference()"></a>二、 aggregateReference()</h1><h6 id="作用：整合参考数据集"><a href="#作用：整合参考数据集" class="headerlink" title="作用：整合参考数据集"></a>作用：整合参考数据集</h6><h6 id="用法"><a href="#用法" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">aggregateReference(ref, labels, power &#x3D; 0.5, assay.type &#x3D; &quot;logcounts&quot;,</span><br><span class="line">check.missing &#x3D; TRUE)</span><br></pre></td></tr></table></figure>
<h6 id="参数："><a href="#参数：" class="headerlink" title="参数："></a>参数：</h6><ul>
<li><strong>ref</strong>：参考表达式值的数字矩阵，通常包含对数表达式值。 或者，包含此类矩阵的SummarizedExperiment对象。<ul>
<li><strong>labels</strong>：参考中所有细胞的已知标记的特征向量或因子。</li>
<li><strong>power</strong>：0到1之间的数字标量，指示每个数据集被整合的权重，请参阅详细信息。</li>
<li><strong>assay.type</strong>：一个整数标量或字符串，如果ref是SummarizedExperiment对象，则指定包含的相关表达式矩阵的ref的数组。</li>
<li><strong>check.missing</strong>：逻辑标量，指示是否应检查行的缺失值（如果找到，则将其删除）。<h6 id="细节"><a href="#细节" class="headerlink" title="细节"></a>细节</h6><blockquote>
<ul>
<li>对于单细胞参考数据集，将单个细胞聚合为伪大量样本以用作参考通常很有用。</li>
</ul>
</blockquote>
</li>
</ul>
</li>
</ul>
<h6 id="例子"><a href="#例子" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">library(scater)</span><br><span class="line">sce &lt;- mockSCE()</span><br><span class="line">sce &lt;- logNormCounts(sce)</span><br><span class="line"># Making up some labels for demonstration purposes:</span><br><span class="line">labels &lt;- sample(LETTERS, ncol(sce), replace&#x3D;TRUE)</span><br><span class="line"># Aggregation at different resolutions:</span><br><span class="line">(aggr &lt;- aggregateReference(sce, labels, power&#x3D;0.5))</span><br><span class="line">(aggr &lt;- aggregateReference(sce, labels, power&#x3D;0))</span><br><span class="line"># No aggregation:</span><br><span class="line">(aggr &lt;- aggregateReference(sce, labels, power&#x3D;1))</span><br></pre></td></tr></table></figure>


<h1 id="三、matchReferences"><a href="#三、matchReferences" class="headerlink" title="三、matchReferences()"></a>三、matchReferences()</h1><h6 id="作用"><a href="#作用" class="headerlink" title="作用"></a>作用</h6><blockquote>
<p>匹配来自两个参考的标签。匹配来自一对引用的标签，这些标签对应于相同的基础细胞类型或状态，但名称有所不同。</p>
</blockquote>
<h6 id="用法-1"><a href="#用法-1" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">matchReferences(ref1, ref2, labels1, labels2, ...)</span><br></pre></td></tr></table></figure>
<h6 id="参数"><a href="#参数" class="headerlink" title="参数"></a>参数</h6><ul>
<li><strong>ref1, ref2</strong>:单细胞（通常是对数转换）表达值的数值矩阵，其中行是基因，列是细胞。 或者，包含此类矩阵的SummarizedExperiment对象</li>
<li><strong>labels1, labels2</strong>:ref1和ref2中所有细胞的已知标记的特征向量或因子。</li>
<li>…：Further arguments to pass to SingleR.<h6 id="返回值"><a href="#返回值" class="headerlink" title="返回值"></a>返回值</h6>包含相互分配的概率表的数字矩阵。 接近1的值表示两个reference之间的标签之间的1：1映射。<h6 id="例子-1"><a href="#例子-1" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">example(SingleR, echo&#x3D;FALSE)</span><br><span class="line">test$label &lt;- paste0(test$label, &quot;_X&quot;) # modifying the labels.</span><br><span class="line">matchReferences(test, ref, labels1&#x3D;test$label, labels2&#x3D;ref$label)</span><br></pre></td></tr></table></figure>

</li>
</ul>
<h1 id="四、-classifySingleR"><a href="#四、-classifySingleR" class="headerlink" title="四、 classifySingleR()"></a>四、 classifySingleR()</h1><h6 id="例子-2"><a href="#例子-2" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ref.se &lt;- BlueprintEncodeData(rm.NA &#x3D; &quot;rows&quot;)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>使用预先训练的包含迭代微调方法的分类器，将标签分配给测试数据集中的每个细胞。(<strong>PS:说人话，就是用来分类细胞的</strong>）</p>
</blockquote>
<h6 id="用法："><a href="#用法：" class="headerlink" title="用法："></a>用法：</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">classifySingleR(test, trained, quantile &#x3D; 0.8, fine.tune &#x3D; TRUE,</span><br><span class="line">tune.thresh &#x3D; 0.05, sd.thresh &#x3D; NULL, prune &#x3D; TRUE,</span><br><span class="line">assay.type &#x3D; &quot;logcounts&quot;, check.missing &#x3D; TRUE,</span><br><span class="line">BPPARAM &#x3D; SerialParam())</span><br></pre></td></tr></table></figure>
<h6 id="参数：-1"><a href="#参数：-1" class="headerlink" title="参数："></a>参数：</h6><ul>
<li><strong>test</strong>： 单细胞表达值的数字矩阵，其中行是基因，列是细胞。或者，包含此类矩阵的SummarizedExperiment对象。</li>
<li><strong>trained</strong>：一个列表，其中包含trainSingleR函数的输出。</li>
<li><strong>quantile</strong>：一个数字标量，指定相关分布的分位数，用于计算每个标签的分值。</li>
<li><strong>fine.tune</strong>：指示是否应执行微调的逻辑标量。</li>
<li><strong>tune.thresh</strong>：一个数字标量，指定与微调中使用的最大相关值之间的最大差。</li>
<li><strong>sd.thresh</strong>：一个数字标量，指定标准偏差的阈值，用于微调过程中的基因选择。   </li>
<li><strong>prune</strong>：逻辑标量，指示是否应执行标签修剪label pruning。</li>
<li><strong>assay.type</strong>：如果test为SummarizedExperiment，则使用整数标量或字符串指定要使用的表达式值矩阵。</li>
<li><strong>check.missing</strong>：逻辑标量，指示是否应检查行的缺失值（如果找到，则将其删除）。</li>
<li><strong>BPPARAM</strong>：BiocParallelParam对象指定要使用的并行化方案。<h6 id="细节："><a href="#细节：" class="headerlink" title="细节："></a>细节：</h6><blockquote>
<ul>
<li><p>仅出于与trainSingleR的一致性，将默认assay.type设置为“ logcounts”。 实际上，也可以使用原始计数（用于UMI数据）或笔录计数（用于读取计数数据），而无需进行标准化和对数转换。 </p>
</li>
<li><p>除了因数值引起的一些细微差异外，任何单调变换都不会影响相关值的计算<br>精确。</p>
</li>
</ul>
</blockquote>
<h6 id="返回值-1"><a href="#返回值-1" class="headerlink" title="返回值"></a>返回值</h6><ul>
<li>每个行都对应于测试集中的细胞的DataFrame。 </li>
</ul>
</li>
</ul>
<ul>
<li>如果<strong>fine.tune = FALSE</strong>，则字段为：<pre><code>- **scores**，是每个细胞（行）中每个标签（列）在指定分位数处的相关性数字矩阵。
- **labels**：一个字符向量，其中包含基于得分最高条目的预测标签。
- **prune.scores**：包含修剪标签的字符向量，其中“低质量”标签被替换为NA。 仅在prune = TRUE时添加。</code></pre></li>
<li>如果<strong>fine.tune = TRUE</strong>，则字段为：<pre><code>- **scores**：同上
- **tuned.scores**：包含第一个和第二个的DataFrame。 这些是在每个细胞的最后一轮微调中包含最佳和次佳分数的数值向量。
- **labels**：微调后包含预测标签的字符向量。
- **prune.scores**：如上所述的修剪标签的字符向量。 仅在prune = TRUE时添加。</code></pre></li>
<li>在<strong>这两种情况</strong>下，DataFrame的元数据都包含：<pre><code>- **common.genes**：在微调之前用于计算相关性的基因的特征向量。
- **de.genes**：用于区分每对标记的基因列表。 仅当在构建训练集时genes =“ de”才返回，有关更多详细信息，请参见 ?trainSingleR。</code></pre><h6 id="例子-3"><a href="#例子-3" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">##############################</span><br><span class="line">## Mocking up training data ##</span><br><span class="line">##############################</span><br><span class="line">Ngroups &lt;- 5</span><br><span class="line">Ngenes &lt;- 1000</span><br><span class="line">means &lt;- matrix(rnorm(Ngenes*Ngroups), nrow&#x3D;Ngenes)</span><br><span class="line">means[1:900,] &lt;- 0</span><br><span class="line">colnames(means) &lt;- LETTERS[1:5]</span><br><span class="line">g &lt;- rep(LETTERS[1:5], each&#x3D;4)</span><br><span class="line">ref &lt;- SummarizedExperiment(</span><br><span class="line">list(counts&#x3D;matrix(rpois(1000*length(g),</span><br><span class="line">lambda&#x3D;10*2^means[,g]), ncol&#x3D;length(g))),</span><br><span class="line">colData&#x3D;DataFrame(label&#x3D;g)</span><br><span class="line">)</span><br><span class="line">rownames(ref) &lt;- sprintf(&quot;GENE_%s&quot;, seq_len(nrow(ref)))</span><br><span class="line">ref &lt;- scater::logNormCounts(ref)</span><br><span class="line">trained &lt;- trainSingleR(ref, ref$label)</span><br><span class="line">###############################</span><br><span class="line">## Mocking up some test data ##</span><br><span class="line">###############################</span><br><span class="line">N &lt;- 100</span><br><span class="line">g &lt;- sample(LETTERS[1:5], N, replace&#x3D;TRUE)</span><br><span class="line">test &lt;- SummarizedExperiment(</span><br><span class="line">list(counts&#x3D;matrix(rpois(1000*N, lambda&#x3D;2^means[,g]), ncol&#x3D;N)),</span><br><span class="line">colData&#x3D;DataFrame(label&#x3D;g)</span><br><span class="line">)</span><br><span class="line">rownames(test) &lt;- sprintf(&quot;GENE_%s&quot;, seq_len(nrow(test)))</span><br><span class="line">test &lt;- scater::logNormCounts(test)</span><br><span class="line">###############################</span><br><span class="line">## Performing classification ##</span><br><span class="line">###############################</span><br><span class="line">pred &lt;- classifySingleR(test, trained)</span><br><span class="line">table(predicted&#x3D;pred$labels, truth&#x3D;g)</span><br></pre></td></tr></table></figure>
<h3 id="五、getDeltaFromMedian"><a href="#五、getDeltaFromMedian" class="headerlink" title="五、getDeltaFromMedian()"></a>五、getDeltaFromMedian()</h3><h6 id="作用-1"><a href="#作用-1" class="headerlink" title="作用"></a>作用</h6>计算每个细胞的增量值，该增量值为：已分配标签的得分与所有标签的和中位数and median 得分之间的差<h6 id="用法-2"><a href="#用法-2" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">getDeltaFromMedian(results)</span><br></pre></td></tr></table></figure>
<h6 id="Arguments"><a href="#Arguments" class="headerlink" title="Arguments"></a>Arguments</h6><ul>
<li><strong>results</strong>: 一个DataFrame，其中包含SingleR或classifySingleR生成的输出。<h6 id="细节-1"><a href="#细节-1" class="headerlink" title="细节"></a>细节</h6>对于想要应用更多自定义过滤器或可视化的用户，此函数计算与pruneScores中使用的相同的增量值。</li>
</ul>
</li>
</ul>
<h6 id="返回值-2"><a href="#返回值-2" class="headerlink" title="返回值"></a>返回值</h6><blockquote>
<p>一个数字向量，其中包含结果中每个细胞的增量值。</p>
</blockquote>
<h6 id="例子-4"><a href="#例子-4" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"># Running the SingleR() example.</span><br><span class="line">example(SingleR, echo&#x3D;FALSE)</span><br><span class="line">summary(getDeltaFromMedian(pred))</span><br></pre></td></tr></table></figure>
<h1 id="六、plotScoreDistribution"><a href="#六、plotScoreDistribution" class="headerlink" title="六、plotScoreDistribution()"></a>六、plotScoreDistribution()</h1><p>绘制标签的分值分布。</p>
<h6 id="用法-3"><a href="#用法-3" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plotScoreDistribution(results, show &#x3D; c(&quot;delta.med&quot;, &quot;delta.next&quot;,</span><br><span class="line">&quot;scores&quot;), labels &#x3D; colnames(results$scores), size &#x3D; 0.5, ncol &#x3D; 5,</span><br><span class="line">dots.on.top &#x3D; TRUE, this.color &#x3D; &quot;#F0E442&quot;,</span><br><span class="line">pruned.color &#x3D; &quot;#E69F00&quot;, other.color &#x3D; &quot;gray60&quot;,</span><br><span class="line">show.nmads &#x3D; NULL, show.min.diff &#x3D; NULL)</span><br></pre></td></tr></table></figure>
<h6 id="参数-1"><a href="#参数-1" class="headerlink" title="参数"></a>参数</h6><ul>
<li><strong>results</strong>：包含SingleR或classifySingleR的输出的DataFrame。</li>
<li><strong>show</strong>：字符串，指定是否显示分打分。与中位数的差还是与下一个最佳分数的差。</li>
<li><strong>labels</strong>：指示要显示的一个或多个标签的字符串向量。 如果为NULL，则显示结果中所有可用的标签。</li>
<li><strong>size</strong>：数字标量，用于设置点的大小。</li>
<li><strong>ncol</strong>：整数标量，用于设置每行显示的标签数。</li>
<li><strong>dots.on.top</strong>:逻辑指定是否在小提琴图的上方绘制细胞点。</li>
<li><strong>this.color</strong>：字符串，用于指定分配给标签的细胞的颜色。</li>
<li><strong>pruned.color</strong>：字符串，用于指定分配给标签但已修剪的细胞的颜色。</li>
<li><strong>other.color</strong>：字符串，用于指定未分配给标签的其他细胞的颜色。</li>
<li><strong>show.nmads</strong>：数字标量，显示使用pruneScores进行修剪的阈值。 仅在show =“ delta.med”时使用。</li>
<li><strong>show.min.diff</strong>：数字标量，显示使用pruneScores进行修剪的阈值。 仅在show =“ delta.med”或“ delta.next”时使用。<h6 id="细节-2"><a href="#细节-2" class="headerlink" title="细节"></a>细节</h6>此函数可创建一个抖动图和小提琴图，以显示一个或多个标签上所有细胞的分配得分或相关值。<h6 id="返回值-3"><a href="#返回值-3" class="headerlink" title="返回值"></a>返回值</h6>ggplot对象，显示小提琴图中的分配的打分。<h6 id="例子-5"><a href="#例子-5" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">example(SingleR, echo&#x3D;FALSE)</span><br><span class="line"># To show the distribution of scores grouped by label:</span><br><span class="line">plotScoreDistribution(results &#x3D; pred)</span><br><span class="line"># We can display a particular label using the label</span><br><span class="line">plotScoreDistribution(results &#x3D; pred, labels &#x3D; &quot;B&quot;)</span><br><span class="line"># To show the distribution of deltas between cells&#39; maximum and median scores,</span><br><span class="line"># grouped by label, change &#96;show&#96; to &quot;delta.med&quot;:</span><br><span class="line"># This is useful for checking&#x2F;adjusting nmads and min.diff.med</span><br><span class="line">plotScoreDistribution(results &#x3D; pred, show &#x3D; &quot;delta.med&quot;)</span><br><span class="line"># The nmads cutoff can be displayed using show.nmads.</span><br><span class="line">plotScoreDistribution(results &#x3D; pred, show &#x3D; &quot;delta.med&quot;,</span><br><span class="line">show.nmads &#x3D; 3)</span><br><span class="line"># A min.diff.med cutoff can be shown using show.min.diff</span><br><span class="line">plotScoreDistribution(results &#x3D; pred, show &#x3D; &quot;delta.med&quot;,</span><br><span class="line">show.min.diff &#x3D; 0.03)</span><br><span class="line"># To show the distribution of deltas between cells&#39; top 2 fine-tuning scores,</span><br><span class="line"># grouped by label, change &#96;show&#96; to &quot;delta.next&quot;:</span><br><span class="line"># This is useful for checking&#x2F;adjusting min.diff.next</span><br><span class="line">plotScoreDistribution(results &#x3D; pred, show &#x3D; &quot;delta.next&quot;)</span><br><span class="line"># A min.diff.med cutoff can be shown using show.min.diff</span><br><span class="line">plotScoreDistribution(results &#x3D; pred, show &#x3D; &quot;delta.next&quot;,</span><br><span class="line">show.min.diff &#x3D; 0.03)</span><br></pre></td></tr></table></figure>

</li>
</ul>
<h1 id="七、plotScoreHeatmap"><a href="#七、plotScoreHeatmap" class="headerlink" title="七、plotScoreHeatmap()"></a>七、plotScoreHeatmap()</h1><h6 id="作用-2"><a href="#作用-2" class="headerlink" title="作用"></a>作用</h6><p>创建所有细胞标签组合的SingleR分配打分的热图。</p>
<h6 id="用法-4"><a href="#用法-4" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">plotScoreHeatmap(results, cells.use &#x3D; NULL, labels.use &#x3D; NULL,</span><br><span class="line">clusters &#x3D; NULL, show.labels &#x3D; FALSE, show.pruned &#x3D; FALSE,</span><br><span class="line">max.labels &#x3D; 40, normalize &#x3D; TRUE, cells.order &#x3D; NULL,</span><br><span class="line">order.by.clusters &#x3D; FALSE, annotation_col &#x3D; NULL,</span><br><span class="line">show_colnames &#x3D; FALSE, ...)</span><br></pre></td></tr></table></figure>
<h6 id="参数-2"><a href="#参数-2" class="headerlink" title="参数"></a>参数</h6><ul>
<li><strong>results</strong>：包含SingleR或classifySingleR的输出的DataFrame。</li>
<li><strong>cells.use</strong>：整数或字符串向量，指定要显示的单个细胞的。 如果为NULL，则显示所有细胞。</li>
<li><strong>labels.use</strong>：指示要显示哪些标签的字符串向量。 如果为NULL，则显示结果中所有可用的标签。</li>
<li><strong>clusters</strong>：包含细胞cluster分配的字符串矢量或因子，将在热图中以注释条显示。</li>
<li><strong>show.labels</strong>：逻辑值，指示是否应将细胞的最终标签显示为注释条。</li>
<li><strong>show.pruned</strong>：逻辑值，指示标签的修剪状态是否应显示为注释条，正如pruneScores所定义。</li>
<li><strong>max.labels</strong>：整数标量，指定要显示的最大标签数。</li>
<li><strong>normalize</strong>：逻辑指定是否应将相关系数归一化为[0，1]。</li>
<li><strong>cells.order</strong>：整数向量，指定热图的细胞/列的顺序。 不论<code>cells.use</code>，此输入的长度应与细胞总数相同。 如果设置，则关闭基于评分的列聚类。</li>
<li><strong>order.by.clusters</strong>：逻辑标量，指定是否应按聚类而不是分数对细胞进行排序。 如果设置，则优先于<code>cells.order</code>输入。</li>
<li><strong>annotation_col, show_colnames, …</strong>：传递给了pheatmap的用于热图控制的其他参数。<br>####### 返回值<br>在当前的图形设备上使用pheatmap生成分配分数的热图。<h6 id="颜色归一化"><a href="#颜色归一化" class="headerlink" title="颜色归一化"></a>颜色归一化</h6>(略）<h6 id="例子-6"><a href="#例子-6" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"># Running the SingleR() example.</span><br><span class="line">example(SingleR, echo&#x3D;FALSE)</span><br><span class="line"># Grab the original identities of the cells as mock clusters</span><br><span class="line">clusts &lt;- g</span><br><span class="line"># Creating a heatmap that shows cells showed.</span><br><span class="line">plotScoreHeatmap(pred)</span><br><span class="line"># Creating a heatmap with clusters displayed.</span><br><span class="line">plotScoreHeatmap(pred, clusters&#x3D;clusts)</span><br><span class="line"># Creating a heatmap with labels displayed.</span><br><span class="line">plotScoreHeatmap(pred, show.labels &#x3D; TRUE)</span><br><span class="line"># Creating a heatmap with whether cells were pruned displayed.</span><br><span class="line">plotScoreHeatmap(pred, show.pruned &#x3D; TRUE)</span><br><span class="line"># We can also turn off the normalization with Normalize &#x3D; FALSE</span><br><span class="line">plotScoreHeatmap(pred, clusters&#x3D;clusts, normalize &#x3D; FALSE)</span><br><span class="line"># To only show certain labels, you can use labels.use or max.labels</span><br><span class="line">plotScoreHeatmap(pred, clusters&#x3D;clusts, labels.use &#x3D; c(&quot;A&quot;,&quot;B&quot;,&quot;D&quot;))</span><br><span class="line">plotScoreHeatmap(pred, clusters&#x3D;clusts, max.labels &#x3D; 4)</span><br><span class="line"># We can pass extra tweaks the heatmap as well</span><br><span class="line">plotScoreHeatmap(pred, clusters&#x3D;clusts, fontsize.row &#x3D; 9)</span><br><span class="line">plotScoreHeatmap(pred, clusters&#x3D;clusts, cutree_col &#x3D; 3)</span><br></pre></td></tr></table></figure>
<h1 id="八、pruneScores"><a href="#八、pruneScores" class="headerlink" title="八、pruneScores()"></a>八、pruneScores()</h1><h6 id="作用-3"><a href="#作用-3" class="headerlink" title="作用"></a>作用</h6>修剪低质量的assignments。删除根据classifySingleR返回的cell-label打分矩阵的低质量assignments。<h6 id="用法-5"><a href="#用法-5" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pruneScores(results, nmads &#x3D; 3, min.diff.med &#x3D; -Inf,</span><br><span class="line">min.diff.next &#x3D; 0, get.thresholds &#x3D; FALSE)</span><br></pre></td></tr></table></figure>
<h6 id="参数-3"><a href="#参数-3" class="headerlink" title="参数"></a>参数</h6></li>
<li><strong>results</strong>:包含着SingleR或classifySingleR生成的输出的DataFrame。</li>
<li><strong>nmads</strong>：数字标量，用于指定在每个标签的增量值分布（即与中位数的差）中定义较低离群值的MAD数值。</li>
<li><strong>min.diff.med</strong>：数字标量，指定每个细胞的最小可接受增量。</li>
<li><strong>min.diff.next</strong>：数字标量，指定微调中最佳分数和次佳分数之间的最小差。</li>
<li><strong>get.thresholds</strong>：逻辑标量，指示是否应返回增量上的每个标签的阈值。<h6 id="细节-3"><a href="#细节-3" class="headerlink" title="细节"></a>细节</h6><blockquote>
<p>就其本身而言，SingleR算法将始终为每个细胞分配标签。即使该细胞的真实标签未在参考标签集中表示，也会发生这种情况，从而导致对该细胞分配了错误的标签。 pruneScores函数旨在通过消除得分较低的低质量分配来减轻这种影响。</p>
</blockquote>
<h6 id="返回值-4"><a href="#返回值-4" class="headerlink" title="返回值"></a>返回值</h6></li>
<li>默认情况下会返回逻辑向量，该逻辑向量指定应忽略结果中的哪些分配。</li>
<li>如果get.thresholds = TRUE，则返回一个数字向量，其中包含增量上每个标签的阈值，这是使用基于异常值的nmads方法定义的。<h6 id="在增量上应用硬过滤器hard-filter"><a href="#在增量上应用硬过滤器hard-filter" class="headerlink" title="在增量上应用硬过滤器hard filter"></a>在增量上应用硬过滤器hard filter</h6></li>
<li>如果指定了min.diff.med，则丢弃增量低于此阈值的细胞。在这种情况下，可以设置最小差异阈值以强制删除低得分细胞。</li>
<li>每个标签的增量分布可能是多峰的，但仍然是正确的。在这种情况下，最好设置nmads = Inf并依靠min.diff.med进行过滤。</li>
<li>请注意，增量不会考虑微调的影响，因为分数在不同的微调步骤之间不具有可比性。<h6 id="基于微调分数过滤"><a href="#基于微调分数过滤" class="headerlink" title="基于微调分数过滤"></a>基于微调分数过滤</h6></li>
<li>如果执行了微调以生成结果，则我们将忽略其微调得分不大于下一个最佳分数min.diff.next的任何细胞。</li>
<li>下一个最 min.diff.next在[0，0.1]之间。 就是说， min.diff.next阈值在涉及高度相关标签的某些应用中可能是有害的。 <h6 id="例子-7"><a href="#例子-7" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><figcaption><span>Running the SingleR() example.</span></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">example(SingleR, echo&#x3D;FALSE)</span><br><span class="line">summary(pruneScores(pred))</span><br><span class="line">pruneScores(pred, get.thresholds&#x3D;TRUE)</span><br><span class="line"># Less stringent:</span><br><span class="line">summary(pruneScores(pred, min.diff.med&#x3D;0))</span><br><span class="line">summary(pruneScores(pred, nmads&#x3D;5))</span><br><span class="line"># More stringent:</span><br><span class="line">summary(pruneScores(pred, min.diff.med&#x3D;0.1))</span><br><span class="line">summary(pruneScores(pred, nmads&#x3D;2))</span><br><span class="line">summary(pruneScores(pred, min.diff.next&#x3D;0.1))</span><br></pre></td></tr></table></figure>
<h1 id="九、-SingleR-）"><a href="#九、-SingleR-）" class="headerlink" title="九、 SingleR(）"></a>九、 SingleR(）</h1><h6 id="作用-4"><a href="#作用-4" class="headerlink" title="作用"></a>作用</h6>注释scRNA-seq数据。给定相同特征空间中的标记参考数据集，返回测试数据集中每个细胞的最佳注释。<h6 id="用法-6"><a href="#用法-6" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">SingleR(test, ref, labels, method &#x3D; c(&quot;single&quot;, &quot;cluster&quot;),</span><br><span class="line">clusters &#x3D; NULL, genes &#x3D; &quot;de&quot;, quantile &#x3D; 0.8, fine.tune &#x3D; TRUE,</span><br><span class="line">tune.thresh &#x3D; 0.05, sd.thresh &#x3D; 1, prune &#x3D; TRUE,</span><br><span class="line">assay.type.test &#x3D; &quot;logcounts&quot;, assay.type.ref &#x3D; &quot;logcounts&quot;,</span><br><span class="line">check.missing &#x3D; TRUE, BNPARAM &#x3D; KmknnParam(),</span><br><span class="line">BPPARAM &#x3D; SerialParam())</span><br></pre></td></tr></table></figure>
<h6 id="参数-4"><a href="#参数-4" class="headerlink" title="参数"></a>参数</h6></li>
<li><strong>test</strong>：single-cell表达值的数字矩阵，其中行是基因，列是细胞。 或者，包含此类矩阵的SummarizedExperiment对象。</li>
<li><strong>ref</strong>： reference expression值的数字矩阵（通常是对数转换，请参阅trainSingleR）。 它应具有与测试中相同的行或其中一部分。或者，包含此类矩阵的SummarizedExperiment对象。</li>
<li><strong>labels</strong>：参考中已知标签的所有样本的字符向量或的因子。</li>
<li><strong>method</strong>：字符串，用于指定是在测试中的单个细胞上执行注释，还是在注释之前将其聚合到群集级配置文件cluster-level profiles中。</li>
<li><strong>clusters</strong>：测试集中每个细胞的字符向量character vector或cluster标识因子。 仅在method =“ cluster”时使用。</li>
<li><strong>genes, sd.thresh</strong>:控制用于注释的基因的参数，请参见trainSingleR。</li>
<li><strong>quantile, fine.tune, tune.thresh, prune</strong>:传递给classifySingleR的更多参数。</li>
<li><strong>assay.type.test</strong>:一个整数标量或字符串，如果测试集是SummarizedExperiment对象，则指定包含相关表达式矩阵的测试集的assay。</li>
<li><strong>assay.type.ref</strong>：一个整数标量或字符串，如果ref是SummarizedExperiment对象，则指定包含相关表达式矩阵的ref的assay。</li>
<li><strong>check.missing</strong>：逻辑标量，指示是否应检查行的缺失值（如果找到，则将其删除）。</li>
<li><strong>BNPARAM</strong>：一个BiocNeighborParam对象，该对象指定用于构建最近邻居索引的算法。</li>
<li><strong>BPPARAM</strong>：BiocParallelParam对象，指定应如何执行并行化（如果有）。<h6 id="细节-4"><a href="#细节-4" class="headerlink" title="细节"></a>细节</h6></li>
<li>如果method =”single”，则此函数实际上只是trainSingleR和classifySingleR的便捷包装。</li>
<li>如果method =”cluster”，则对per-cell profiles求和以获得 per-cluster profiles，并在这些cluster上执行注释。</li>
<li>该功能将自动将分析限制为参考和测试中可用的基因的交集。 如果此交集为空（例如，因为两个数据集在其行名中使用不同的注释），则会引发错误。</li>
<li>ref可以包含单个细胞数据或bulk数据，但对于前者，请阅读?trainSingleR中的注释。<h6 id="返回值-5"><a href="#返回值-5" class="headerlink" title="返回值"></a>返回值</h6>返回一个DataFrame，其中包含每个细胞或cluster（行）的注释统计信息。 这与classifySingleR的输出相同。<br>####### 例子<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">##############################</span><br><span class="line">## Mocking up training data ##</span><br><span class="line">##############################</span><br><span class="line">Ngroups &lt;- 5</span><br><span class="line">Ngenes &lt;- 1000</span><br><span class="line">means &lt;- matrix(rnorm(Ngenes*Ngroups), nrow&#x3D;Ngenes)</span><br><span class="line">means[1:900,] &lt;- 0</span><br><span class="line">colnames(means) &lt;- LETTERS[1:5]</span><br><span class="line">g &lt;- rep(LETTERS[1:5], each&#x3D;4)</span><br><span class="line">ref &lt;- SummarizedExperiment(</span><br><span class="line">list(counts&#x3D;matrix(rpois(1000*length(g),</span><br><span class="line">lambda&#x3D;10*2^means[,g]), ncol&#x3D;length(g))),</span><br><span class="line">colData&#x3D;DataFrame(label&#x3D;g)</span><br><span class="line">)</span><br><span class="line">rownames(ref) &lt;- sprintf(&quot;GENE_%s&quot;, seq_len(nrow(ref)))</span><br><span class="line">ref &lt;- scater::logNormCounts(ref)</span><br><span class="line">trained &lt;- trainSingleR(ref, ref$label)</span><br><span class="line">###############################</span><br><span class="line">## Mocking up some test data ##</span><br><span class="line">###############################</span><br><span class="line">N &lt;- 100</span><br><span class="line">g &lt;- sample(LETTERS[1:5], N, replace&#x3D;TRUE)</span><br><span class="line">test &lt;- SummarizedExperiment(</span><br><span class="line">list(counts&#x3D;matrix(rpois(1000*N, lambda&#x3D;2^means[,g]), ncol&#x3D;N)),</span><br><span class="line">colData&#x3D;DataFrame(label&#x3D;g)</span><br><span class="line">)</span><br><span class="line">rownames(test) &lt;- sprintf(&quot;GENE_%s&quot;, seq_len(nrow(test)))</span><br><span class="line">test &lt;- scater::logNormCounts(test)</span><br><span class="line">###############################</span><br><span class="line">## Performing classification ##</span><br><span class="line">###############################</span><br><span class="line">pred &lt;- SingleR(test, ref, labels&#x3D;ref$label)</span><br><span class="line">table(predicted&#x3D;pred$labels, truth&#x3D;g)</span><br><span class="line">pred2 &lt;- SingleR(test, ref, labels&#x3D;ref$label,</span><br><span class="line">method&#x3D;&quot;cluster&quot;, clusters&#x3D;test$label)</span><br><span class="line">table(predicted&#x3D;pred2$labels, truth&#x3D;rownames(pred2))</span><br></pre></td></tr></table></figure>
<h1 id="十、trainSingleR"><a href="#十、trainSingleR" class="headerlink" title="十、trainSingleR"></a>十、trainSingleR</h1><h6 id="作用-5"><a href="#作用-5" class="headerlink" title="作用"></a>作用</h6>在具有已知标签的参考数据集上训练SingleR分类器。<h6 id="用法-7"><a href="#用法-7" class="headerlink" title="用法"></a>用法</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">trainSingleR(ref, labels, genes &#x3D; &quot;de&quot;, sd.thresh &#x3D; 1, de.n &#x3D; NULL,</span><br><span class="line">assay.type &#x3D; &quot;logcounts&quot;, check.missing &#x3D; TRUE,</span><br><span class="line">BNPARAM &#x3D; KmknnParam())</span><br></pre></td></tr></table></figure>
<h6 id="参数-5"><a href="#参数-5" class="headerlink" title="参数"></a>参数</h6></li>
<li><strong>ref</strong>：表达值的数字矩阵，其中行是基因，列是参考样本（单个细胞或bulk样本）。 这些 expression values应进行对数转换，请参阅“详细信息”。 或者，包含此类矩阵的SummarizedExperiment对象。</li>
<li><strong>labels</strong>：参考中所有样本的字符向量或已知标签的因子。</li>
<li><strong>genes</strong>：一个字符串，指定要使用的特征feature选择方法，请参见“详细信息”。可选地，在标签对pairs of labels之间包含DE基因的字符向量的列表。或者，包含每个标签的marker基因的字符向量列表。</li>
<li>*<em>sd.thresh *</em>:一个数字标量，指定每个基因的标准偏差的最小阈值。 仅在genes =”sd”时使用。</li>
<li><strong>de.n</strong>：一个整数标量，指定genes =”de”时要使用的DE基因的数量。 默认值为500 *（2/3）^ log2（N），其中N是唯一标签的数量。</li>
<li><strong>assay.type</strong>:一个整数标量或字符串，如果ref是SummarizedExperiment对象，则指定包含相关表达式矩阵的ref的assay。</li>
<li><strong>check.missing</strong>: 逻辑标量，指示是否应检查行的缺失值（如果找到，则将其删除）。</li>
<li><strong>BNPARAM</strong>：一个BiocNeighborParam对象，该对象指定用于构建最近邻居索引的算法。<h6 id="细节-5"><a href="#细节-5" class="headerlink" title="细节"></a>细节</h6>几个选项可用于功能选择：</li>
<li>genes=”de”：识别在标签之间差异表达的基因</li>
<li>genes=”sd：识别在标签之间变化很大的基因。</li>
<li>genes=”all”：将不会执行任何功能选择。<br>如果genes =”de”或”sd”，则 expression values 应是对数转换和归一化后的。<br>或者，基因可以是字符向量的命名列表：<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">genes &lt;- list(</span><br><span class="line">A &#x3D; list(A &#x3D; character(0), B &#x3D; &quot;GENE_1&quot;, C &#x3D; c(&quot;GENE_2&quot;, &quot;GENE_3&quot;)),</span><br><span class="line">B &#x3D; list(A &#x3D; &quot;GENE_100&quot;, B &#x3D; character(0), C &#x3D; &quot;GENE_200&quot;),</span><br><span class="line">C &#x3D; list(A &#x3D; c(&quot;GENE_4&quot;, &quot;GENE_5&quot;), B &#x3D; &quot;GENE_5&quot;, C &#x3D; character(0))</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
如果我们考虑条目基因$A $B，它包含标签“ A”相对于标签“ B”的marker基因。 即，与“ B”相比，这些基因在“ A”中被上调。 每个标签的外部列表应具有一个列表，每个标签的内部列表应具有一个字符向量。 （显然，标签本身不能带有标记markers，因此将其设置为character(0)。）</li>
</ul>
<p>或者，基因可以是字符向量的命名列表：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">genes &lt;- list(</span><br><span class="line">A &#x3D; c(&quot;GENE_1&quot;, &quot;GENE_2&quot;, &quot;GENE_3&quot;),</span><br><span class="line">B &#x3D; c(&quot;GENE_100&quot;, &quot;GENE_200&quot;),</span><br><span class="line">C &#x3D; c(&quot;GENE_4&quot;, &quot;GENE_5&quot;)</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<h6 id="返回值-6"><a href="#返回值-6" class="headerlink" title="返回值"></a>返回值</h6><p>包含以下内容的列表：</p>
<ul>
<li>common.genes：通过feature selection方法选择的所有基因的字符向量</li>
<li>nn.indices：BiocNeighborIndex对象的列表，其中包含预先构造的邻居搜索索引。</li>
<li>original.exprs：数字矩阵的列表，其中每个矩阵都包含特定标签的所有细胞。</li>
<li>extra：有关 feature selection的其他信息的列表，供classifySingleR使用。<br>这包括基因中的选择方法以及在分类过程中可以重复使用的方法特定结构。<h6 id="关于单细胞参考的注意事项"><a href="#关于单细胞参考的注意事项" class="headerlink" title="关于单细胞参考的注意事项"></a>关于单细胞参考的注意事项</h6></li>
<li>默认marker选择基于中位数的对数倍变化，并且在设计时要考虑bulk参考。 对于单细胞参考数据来说，零计数超过50％的情况并不罕见，这方法可能并不有效。</li>
<li>建议用户要么汇总其单细胞参考以创建pseudo-bulk samples，要么在外部检测markers并将markers列表传递给基因（请参见示例）<h6 id="例子-8"><a href="#例子-8" class="headerlink" title="例子"></a>例子</h6><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">##############################</span><br><span class="line">## Mocking up training data ##</span><br><span class="line">##############################</span><br><span class="line">Ngroups &lt;- 5</span><br><span class="line">Ngenes &lt;- 1000</span><br><span class="line">means &lt;- matrix(rnorm(Ngenes*Ngroups), nrow&#x3D;Ngenes)</span><br><span class="line">means[1:900,] &lt;- 0</span><br><span class="line">colnames(means) &lt;- LETTERS[1:5]</span><br><span class="line">g &lt;- rep(LETTERS[1:5], each&#x3D;4)</span><br><span class="line">ref &lt;- SummarizedExperiment(</span><br><span class="line">list(counts&#x3D;matrix(rpois(1000*length(g),</span><br><span class="line">lambda&#x3D;10*2^means[,g]), ncol&#x3D;length(g))),</span><br><span class="line">colData&#x3D;DataFrame(label&#x3D;g)</span><br><span class="line">)</span><br><span class="line">rownames(ref) &lt;- sprintf(&quot;GENE_%s&quot;, seq_len(nrow(ref)))</span><br><span class="line">########################</span><br><span class="line">## Doing the training ##</span><br><span class="line">########################</span><br><span class="line"># Normalizing and log-transforming for automated marker detection.</span><br><span class="line">ref &lt;- scater::logNormCounts(ref)</span><br><span class="line">trained &lt;- trainSingleR(ref, ref$label)</span><br><span class="line">trained</span><br><span class="line">trained$nn.indices</span><br><span class="line">length(trained$common.genes)</span><br><span class="line"># Alternatively, computing and supplying a set of label-specific markers.</span><br><span class="line">by.t &lt;- scran::pairwiseTTests(assay(ref, 2), ref$label)</span><br><span class="line">markers &lt;- scran::getTopMarkers(by.t[[1]], by.t[[2]], n&#x3D;10)</span><br><span class="line">trained &lt;- trainSingleR(ref, ref$label, genes&#x3D;markers)</span><br><span class="line">length(trained$common.genes)</span><br></pre></td></tr></table></figure>
</li>
</ul>

        </div>
        
        
        
    </div>
</div>



<div class="card">
    <div class="card-content">
        <h3 class="menu-label has-text-centered">喜欢这篇文章？打赏一下作者吧</h3>
        <div class="buttons is-centered">
            
                
<div class="notification is-danger">
    You forgot to set the <code>qrcode</code> for Alipay. Please set it in <code>_config.yml</code>.
</div>

                
                
<div class="notification is-danger">
    You forgot to set the <code>qrcode</code> for Wechat. Please set it in <code>_config.yml</code>.
</div>

                
                <!-- Visit https://www.paypal.com/donate/buttons/ to get your donate button -->

<div class="notification is-danger">
    You forgot to set the <code>business</code> and <code>currency_code</code> for Paypal. Please set it in <code>_config.yml</code>.
</div>

                
                
<div class="notification is-danger">
    You forgot to set the <code>url</code> Patreon. Please set it in <code>_config.yml</code>.
</div>

                
        </div>
    </div>
</div>



<div class="card card-transparent">
    <div class="level post-navigation is-flex-wrap is-mobile">
        
        <div class="level-start">
            <a class="level level-item has-link-grey  article-nav-prev" href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%8415%E3%80%91%E6%97%A7%E7%89%88SingleR-1-0-0%E7%9A%84%E5%9F%BA%E6%9C%AC%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95/">
                <i class="level-item fas fa-chevron-left"></i>
                <span class="level-item">【单细胞转录组15】旧版SingleR 1.0.0的基本使用方法</span>
            </a>
        </div>
        
        
        <div class="level-end">
            <a class="level level-item has-link-grey  article-nav-next" href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%849%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E7%9A%84%E4%BB%A3%E7%A0%81%E5%AE%9E%E6%93%8D/">
                <span class="level-item">【单细胞转录组9】新版SingleR1.1.1细胞注释的代码实操</span>
                <i class="level-item fas fa-chevron-right"></i>
            </a>
        </div>
        
    </div>
</div>


</div>
                




<div class="column is-4-tablet is-4-desktop is-3-widescreen  has-order-1 column-left ">
    
        
<div class="card widget">
    <div class="card-content">
        <nav class="level">
            <div class="level-item has-text-centered" style="flex-shrink: 1">
                <div>
                    
                    <figure class="image is-128x128 has-mb-6">
                        <img class="" src="/images/avatar.png" alt="Robin唔系肉饼">
                    </figure>
                    
                    <p class="is-size-4 is-block">
                        Robin唔系肉饼
                    </p>
                    
                    
                    <p class="is-size-6 is-block">
                        湾区青年，广佛候鸟
                    </p>
                    
                    
                    <p class="is-size-6 is-flex is-flex-center has-text-grey">
                        <i class="fas fa-map-marker-alt has-mr-7"></i>
                        <span>Guangdong, China</span>
                    </p>
                    
                </div>
            </div>
        </nav>
        <nav class="level is-mobile">
            <div class="level-item has-text-centered is-marginless">
                <div>
                    <p class="heading">
                        文章
                    </p>
                    <a href="/archives">
                        <p class="title has-text-weight-normal">
                            7
                        </p>
                    </a>
                </div>
            </div>
            <div class="level-item has-text-centered is-marginless">
                <div>
                    <p class="heading">
                        分类
                    </p>
                    <a href="/categories">
                        <p class="title has-text-weight-normal">
                            0
                        </p>
                    </a>
                </div>
            </div>
            <div class="level-item has-text-centered is-marginless">
                <div>
                    <p class="heading">
                        标签
                    </p>
                    <a href="/tags">
                        <p class="title has-text-weight-normal">
                            1
                        </p>
                    </a>
                </div>
            </div>
        </nav>
        
        <div class="level">
            <a class="level-item button is-link is-rounded" href="https://github.com/BioAIEvolu" target="_blank" rel="noopener">
                关注我</a>
        </div>
        
        
        
        <div class="level is-mobile">
            
            <a class="level-item button is-white is-marginless" target="_blank" rel="noopener"
                title="Github" href="https://github.com/BioAIEvolu">
                
                <i class="fab fa-github"></i>
                
            </a>
            
            <a class="level-item button is-white is-marginless" target="_blank" rel="noopener"
                title="Facebook" href="https://facebook.com">
                
                <i class="fab fa-facebook"></i>
                
            </a>
            
            <a class="level-item button is-white is-marginless" target="_blank" rel="noopener"
                title="Twitter" href="https://twitter.com">
                
                <i class="fab fa-twitter"></i>
                
            </a>
            
            <a class="level-item button is-white is-marginless" target="_blank" rel="noopener"
                title="Dribbble" href="https://dribbble.com">
                
                <i class="fab fa-dribbble"></i>
                
            </a>
            
            <a class="level-item button is-white is-marginless" target="_blank" rel="noopener"
                title="RSS" href="/">
                
                <i class="fas fa-rss"></i>
                
            </a>
            
        </div>
        
    </div>
</div>
    
        
    
        <div class="card widget">
    <div class="card-content">
        <div class="menu">
        <h3 class="menu-label">
            链接
        </h3>
        <ul class="menu-list">
        
            <li>
                <a class="level is-mobile" href="https://hexo.io" target="_blank" rel="noopener">
                    <span class="level-left">
                        <span class="level-item">Hexo</span>
                    </span>
                    <span class="level-right">
                        <span class="level-item tag">hexo.io</span>
                    </span>
                </a>
            </li>
        
        </ul>
        </div>
    </div>
</div>

    
        
<div class="card widget">
    <div class="card-content">
        <div class="menu">
            <h3 class="menu-label">
                分类
            </h3>
            <ul class="menu-list">
            
            </ul>
        </div>
    </div>
</div>
    
        <div class="card widget">
    <div class="card-content">
        <h3 class="menu-label">
            标签云
        </h3>
        <a href="/tags/%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80/" style="font-size: 10px;"> "生信基础"</a>
    </div>
</div>
    
    
        <div class="column-right-shadow is-hidden-widescreen ">
        
            <div class="card widget">
    <div class="card-content">
        <h3 class="menu-label">
            最新文章
        </h3>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%801%E3%80%91%E6%A0%B9%E6%8D%AESRA%E5%8F%B7%E6%89%B9%E9%87%8F%E4%B8%8B%E8%BD%BDfastq%E6%96%87%E4%BB%B6%E7%9A%84%E4%B8%80%E7%A7%8D%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【生信基础1】根据SRA号批量下载fastq文件的一种实现方法">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T13:54:03.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%801%E3%80%91%E6%A0%B9%E6%8D%AESRA%E5%8F%B7%E6%89%B9%E9%87%8F%E4%B8%8B%E8%BD%BDfastq%E6%96%87%E4%BB%B6%E7%9A%84%E4%B8%80%E7%A7%8D%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【生信基础1】根据SRA号批量下载fastq文件的一种实现方法</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%8415%E3%80%91%E6%97%A7%E7%89%88SingleR-1-0-0%E7%9A%84%E5%9F%BA%E6%9C%AC%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【单细胞转录组15】旧版SingleR 1.0.0的基本使用方法">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T08:19:42.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%8415%E3%80%91%E6%97%A7%E7%89%88SingleR-1-0-0%E7%9A%84%E5%9F%BA%E6%9C%AC%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【单细胞转录组15】旧版SingleR 1.0.0的基本使用方法</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%848%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E5%8C%85%E7%9A%84%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T08:17:54.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%848%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E5%8C%85%E7%9A%84%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%849%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E7%9A%84%E4%BB%A3%E7%A0%81%E5%AE%9E%E6%93%8D/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【单细胞转录组9】新版SingleR1.1.1细胞注释的代码实操">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T08:09:19.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%849%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E7%9A%84%E4%BB%A3%E7%A0%81%E5%AE%9E%E6%93%8D/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【单细胞转录组9】新版SingleR1.1.1细胞注释的代码实操</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E7%88%AC%E8%99%AB%E9%A1%B9%E7%9B%AE2%E3%80%91Selenium%E6%A1%86%E6%9E%B6%E7%88%AC%E5%8F%96%E6%B7%98%E5%AE%9D%E5%95%86%E5%93%81%E4%BF%A1%E6%81%AF/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【爬虫项目(二)】Selenium框架爬取淘宝商品信息">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T07:44:27.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E7%88%AC%E8%99%AB%E9%A1%B9%E7%9B%AE2%E3%80%91Selenium%E6%A1%86%E6%9E%B6%E7%88%AC%E5%8F%96%E6%B7%98%E5%AE%9D%E5%95%86%E5%93%81%E4%BF%A1%E6%81%AF/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【爬虫项目(二)】Selenium框架爬取淘宝商品信息</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
    </div>
</div>
        
            <div class="card widget">
    <div class="card-content">
        <div class="menu">
        <h3 class="menu-label">
            归档
        </h3>
        <ul class="menu-list">
        
        <li>
            <a class="level is-marginless" href="/archives/2020/02/">
                <span class="level-start">
                    <span class="level-item">二月 2020</span>
                </span>
                <span class="level-end">
                    <span class="level-item tag">7</span>
                </span>
            </a>
        </li>
        
        </ul>
        </div>
    </div>
</div>
        
            <div class="card widget">
    <div class="card-content">
        <div class="menu">
            <h3 class="menu-label">
                标签
            </h3>
            <div class="field is-grouped is-grouped-multiline">
                
                <div class="control">
                    <a class="tags has-addons" href="/tags/%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80/">
                        <span class="tag"> &#34;生信基础&#34;</span>
                        <span class="tag is-grey">1</span>
                    </a>
                </div>
                
            </div>
        </div>
    </div>
</div>
        
        </div>
    
</div>

                




<div class="column is-4-tablet is-4-desktop is-3-widescreen is-hidden-touch is-hidden-desktop-only has-order-3 column-right ">
    
        <div class="card widget">
    <div class="card-content">
        <h3 class="menu-label">
            最新文章
        </h3>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%801%E3%80%91%E6%A0%B9%E6%8D%AESRA%E5%8F%B7%E6%89%B9%E9%87%8F%E4%B8%8B%E8%BD%BDfastq%E6%96%87%E4%BB%B6%E7%9A%84%E4%B8%80%E7%A7%8D%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【生信基础1】根据SRA号批量下载fastq文件的一种实现方法">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T13:54:03.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%801%E3%80%91%E6%A0%B9%E6%8D%AESRA%E5%8F%B7%E6%89%B9%E9%87%8F%E4%B8%8B%E8%BD%BDfastq%E6%96%87%E4%BB%B6%E7%9A%84%E4%B8%80%E7%A7%8D%E5%AE%9E%E7%8E%B0%E6%96%B9%E6%B3%95/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【生信基础1】根据SRA号批量下载fastq文件的一种实现方法</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%8415%E3%80%91%E6%97%A7%E7%89%88SingleR-1-0-0%E7%9A%84%E5%9F%BA%E6%9C%AC%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【单细胞转录组15】旧版SingleR 1.0.0的基本使用方法">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T08:19:42.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%8415%E3%80%91%E6%97%A7%E7%89%88SingleR-1-0-0%E7%9A%84%E5%9F%BA%E6%9C%AC%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【单细胞转录组15】旧版SingleR 1.0.0的基本使用方法</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%848%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E5%8C%85%E7%9A%84%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T08:17:54.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%848%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E5%8C%85%E7%9A%84%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%849%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E7%9A%84%E4%BB%A3%E7%A0%81%E5%AE%9E%E6%93%8D/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【单细胞转录组9】新版SingleR1.1.1细胞注释的代码实操">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T08:09:19.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%849%E3%80%91%E6%96%B0%E7%89%88SingleR1-1-1%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A%E7%9A%84%E4%BB%A3%E7%A0%81%E5%AE%9E%E6%93%8D/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【单细胞转录组9】新版SingleR1.1.1细胞注释的代码实操</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
        <article class="media">
            
            <a href="/2020/02/06/%E3%80%90%E7%88%AC%E8%99%AB%E9%A1%B9%E7%9B%AE2%E3%80%91Selenium%E6%A1%86%E6%9E%B6%E7%88%AC%E5%8F%96%E6%B7%98%E5%AE%9D%E5%95%86%E5%93%81%E4%BF%A1%E6%81%AF/" class="media-left">
                <p class="image is-64x64">
                    <img class="thumbnail" src="/images/thumbnail.svg" alt="【爬虫项目(二)】Selenium框架爬取淘宝商品信息">
                </p>
            </a>
            
            <div class="media-content">
                <div class="content">
                    <div><time class="has-text-grey is-size-7 is-uppercase" datetime="2020-02-06T07:44:27.000Z">2020-02-06</time></div>
                    <a href="/2020/02/06/%E3%80%90%E7%88%AC%E8%99%AB%E9%A1%B9%E7%9B%AE2%E3%80%91Selenium%E6%A1%86%E6%9E%B6%E7%88%AC%E5%8F%96%E6%B7%98%E5%AE%9D%E5%95%86%E5%93%81%E4%BF%A1%E6%81%AF/" class="title has-link-black-ter is-size-6 has-text-weight-normal">【爬虫项目(二)】Selenium框架爬取淘宝商品信息</a>
                    <p class="is-size-7 is-uppercase">
                        
                    </p>
                </div>
            </div>
        </article>
        
    </div>
</div>
    
        <div class="card widget">
    <div class="card-content">
        <div class="menu">
        <h3 class="menu-label">
            归档
        </h3>
        <ul class="menu-list">
        
        <li>
            <a class="level is-marginless" href="/archives/2020/02/">
                <span class="level-start">
                    <span class="level-item">二月 2020</span>
                </span>
                <span class="level-end">
                    <span class="level-item tag">7</span>
                </span>
            </a>
        </li>
        
        </ul>
        </div>
    </div>
</div>
    
        <div class="card widget">
    <div class="card-content">
        <div class="menu">
            <h3 class="menu-label">
                标签
            </h3>
            <div class="field is-grouped is-grouped-multiline">
                
                <div class="control">
                    <a class="tags has-addons" href="/tags/%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80/">
                        <span class="tag"> &#34;生信基础&#34;</span>
                        <span class="tag is-grey">1</span>
                    </a>
                </div>
                
            </div>
        </div>
    </div>
</div>
    
    
</div>

            </div>
        </div>
    </section>
    <footer class="footer">
    <div class="container">
        <div class="level">
            <div class="level-start has-text-centered-mobile">
                <a class="footer-logo is-block has-mb-6" href="/">
                
                    <img src="/images/logo.svg" alt="【单细胞转录组8】新版SingleR1.1.1细胞注释包的学习笔记" height="28">
                
                </a>
                <p class="is-size-7">
                &copy; 2020 Robin Hood&nbsp;
                Powered by <a href="https://hexo.io/" target="_blank" rel="noopener">Hexo</a> & <a
                        href="https://github.com/ppoffice/hexo-theme-icarus" target="_blank" rel="noopener">Icarus</a>
                
                </p>
            </div>
            <div class="level-end">
            
                <div class="field has-addons is-flex-center-mobile has-mt-5-mobile is-flex-wrap is-flex-middle">
                
                <p class="control">
                    <a class="button is-white is-large" target="_blank" rel="noopener" title="Creative Commons" href="https://creativecommons.org/">
                        
                        <i class="fab fa-creative-commons"></i>
                        
                    </a>
                </p>
                
                <p class="control">
                    <a class="button is-white is-large" target="_blank" rel="noopener" title="Attribution 4.0 International" href="https://creativecommons.org/licenses/by/4.0/">
                        
                        <i class="fab fa-creative-commons-by"></i>
                        
                    </a>
                </p>
                
                <p class="control">
                    <a class="button is-white is-large" target="_blank" rel="noopener" title="Download on GitHub" href="https://github.com/ppoffice/hexo-theme-icarus">
                        
                        <i class="fab fa-github"></i>
                        
                    </a>
                </p>
                
                </div>
            
            </div>
        </div>
    </div>
</footer>
    <script src="https://cdn.jsdelivr.net/npm/jquery@3.3.1/dist/jquery.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/moment@2.22.2/min/moment-with-locales.min.js"></script>
<script>moment.locale("zh-CN");</script>


<script>
var IcarusThemeSettings = {
    site: {
        url: 'http://yoursite.com',
        external_link: {"enable":true,"exclude":[]}
    },
    article: {
        highlight: {
            clipboard: true,
            fold: 'unfolded'
        }
    }
};
</script>


<script src="https://cdn.jsdelivr.net/npm/clipboard@2.0.4/dist/clipboard.min.js" defer></script>





<script src="/js/animation.js"></script>



<script src="https://cdn.jsdelivr.net/npm/lightgallery@1.6.8/dist/js/lightgallery.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/justifiedGallery@3.7.0/dist/js/jquery.justifiedGallery.min.js" defer></script>
<script src="/js/gallery.js" defer></script>



<div id="outdated">
    <h6>Your browser is out-of-date!</h6>
    <p>Update your browser to view this website correctly. <a id="btnUpdateBrowser" href="http://outdatedbrowser.com/">Update
            my browser now </a></p>
    <p class="last"><a href="#" id="btnCloseUpdateBrowser" title="Close">&times;</a></p>
</div>
<script src="https://cdn.jsdelivr.net/npm/outdatedbrowser@1.1.5/outdatedbrowser/outdatedbrowser.min.js" defer></script>
<script>
    document.addEventListener("DOMContentLoaded", function () {
        outdatedBrowser({
            bgColor: '#f25648',
            color: '#ffffff',
            lowerThan: 'flex'
        });
    });
</script>


<script src="https://cdn.jsdelivr.net/npm/mathjax@2.7.5/unpacked/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
<script>
document.addEventListener('DOMContentLoaded', function () {
    MathJax.Hub.Config({
        'HTML-CSS': {
            matchFontHeight: false
        },
        SVG: {
            matchFontHeight: false
        },
        CommonHTML: {
            matchFontHeight: false
        },
        tex2jax: {
            inlineMath: [
                ['$','$'],
                ['\\(','\\)']
            ]
        }
    });
});
</script>


<a id="back-to-top" title="回到顶端" href="javascript:;">
    <i class="fas fa-chevron-up"></i>
</a>
<script src="/js/back-to-top.js" defer></script>














<script src="/js/main.js" defer></script>

    
    <div class="searchbox ins-search">
    <div class="searchbox-container ins-search-container">
        <div class="searchbox-input-wrapper">
            <input type="text" class="searchbox-input ins-search-input" placeholder="想要查找什么..." />
            <span class="searchbox-close ins-close ins-selectable"><i class="fa fa-times-circle"></i></span>
        </div>
        <div class="searchbox-result-wrapper ins-section-wrapper">
            <div class="ins-section-container"></div>
        </div>
    </div>
</div>
<script>
    (function (window) {
        var INSIGHT_CONFIG = {
            TRANSLATION: {
                POSTS: '文章',
                PAGES: '页面',
                CATEGORIES: '分类',
                TAGS: '标签',
                UNTITLED: '(无标题)',
            },
            CONTENT_URL: '/content.json',
        };
        window.INSIGHT_CONFIG = INSIGHT_CONFIG;
    })(window);
</script>
<script src="/js/insight.js" defer></script>
<link rel="stylesheet" href="/css/search.css">
<link rel="stylesheet" href="/css/insight.css">
    
</body>
</html>